src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1137 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::command::AuthenticatorClientPinParameters;
+use super::data_formats::{ClientPinSubCommand, CoseKey, GetAssertionHmacSecretInput};
+use super::response::{AuthenticatorClientPinResponse, ResponseData};
+use super::status_code::Ctap2StatusCode;
+use super::storage::PersistentStore;
+#[cfg(feature = "with_ctap2_1")]
+use alloc::string::String;
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use crypto::cbc::{cbc_decrypt, cbc_encrypt};
+use crypto::hmac::{hmac_256, verify_hmac_256_first_128bits};
+use crypto::rng256::Rng256;
+use crypto::sha256::Sha256;
+use crypto::Hash256;
+#[cfg(all(test, feature = "with_ctap2_1"))]
+use enum_iterator::IntoEnumIterator;
+use subtle::ConstantTimeEq;
+
+// Those constants have to be multiples of 16, the AES block size.
+pub const PIN_AUTH_LENGTH: usize = 16;
+const PIN_PADDED_LENGTH: usize = 64;
+const PIN_TOKEN_LENGTH: usize = 32;
+
+/// Checks the given pin_auth against the truncated output of HMAC-SHA256.
+fn check_pin_auth(hmac_key: &[u8], hmac_contents: &[u8], pin_auth: &[u8]) -> bool {
+    if pin_auth.len() != PIN_AUTH_LENGTH {
+        return false;
+    }
+    verify_hmac_256_first_128bits::<Sha256>(
+        hmac_key,
+        hmac_contents,
+        array_ref![pin_auth, 0, PIN_AUTH_LENGTH],
+    )
+}
+
+/// Decrypts the HMAC secret salt(s) that were encrypted with the shared secret.
+/// The credRandom is used as a secret to HMAC those salts.
+/// The last step is to re-encrypt the outputs.
+fn encrypt_hmac_secret_output(
+    shared_secret: &[u8; 32],
+    salt_enc: &[u8],
+    cred_random: &[u8],
+) -> Result<Vec<u8>, Ctap2StatusCode> {
+    if salt_enc.len() != 32 && salt_enc.len() != 64 {
+        return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+    }
+    if cred_random.len() != 32 {
+        // We are strict here. We need at least 32 byte, but expect exactly 32.
+        return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+    }
+    let aes_enc_key = crypto::aes256::EncryptionKey::new(shared_secret);
+    let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+    // The specification specifically asks for a zero IV.
+    let iv = [0u8; 16];
+
+    let mut cred_random_secret = [0u8; 32];
+    cred_random_secret.clone_from_slice(cred_random);
+
+    // Initialization of 4 blocks in any case makes this function more readable.
+    let mut blocks = [[0u8; 16]; 4];
+    let block_len = salt_enc.len() / 16;
```

Comment:
What happens if the length of `salt_enc` is not a multiple of 16? Might be worth documenting the expected length of this slice.

Reply:
It's documented in line 58:
`if salt_enc.len() != 32 && salt_enc.len() != 64 {`
`    return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);`
`}`

I'l add an additional comment.